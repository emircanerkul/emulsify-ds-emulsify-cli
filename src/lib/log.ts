/**
 * @file log.ts
 * Exports methods that MUST be used when writing to the console.
 */

import { Chalk, dim, cyan, yellow, green, white } from 'chalk';
import R from 'ramda';

export type LogMethod =
  | 'info'
  | 'error'
  | 'warn'
  | 'debug'
  | 'verbose'
  | 'success';

const logMethodColorMap: {
  [name in LogMethod]: Chalk;
} = {
  info: cyan,
  error: white.bgRed.bold,
  warn: yellow.bold,
  debug: dim,
  verbose: dim,
  success: green,
};

const withColor = (
  logger: Console['log'] | Console['info'] | Console['error']
) => (method: LogMethod, message: string): void =>
  /* eslint-disable-next-line security/detect-object-injection */
  logger(logMethodColorMap[method](message));

const logMethodEq = (potentialMethod: LogMethod) => (
  method: LogMethod,
  _: string
): boolean => potentialMethod === method;

/**
 * Lib function that allows for info, error, warn, debug, verbose, and success messages
 * to be written to the console with consistent methods and colors.
 *
 * @param message string containing message to be logged.
 * @param method method that should be used to log, such as 'error', or 'warn'.
 */
export default function log(method: LogMethod, message: string): void {
  // @TODO: add support for --verbose flag, and suppress verbose messages
  // by default when --verbose is false or void.

  // emit log message based off of method.
  R.cond([
    [logMethodEq('error'), withColor(console.error)],
    [logMethodEq('info'), withColor(console.info)],
    [logMethodEq('warn'), withColor(console.warn)],
    [R.T, withColor(console.log)],
  ])(method, message);

  return;
}
