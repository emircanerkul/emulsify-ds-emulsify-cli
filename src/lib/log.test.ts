import R from 'ramda';

jest.spyOn(global.console, 'log').mockImplementation(R.identity);
jest.spyOn(global.console, 'info').mockImplementation(R.identity);
jest.spyOn(global.console, 'error').mockImplementation(R.identity);
jest.spyOn(global.console, 'warn').mockImplementation(R.identity);

import log from './log';

describe('log', () => {
  it('can log info messages', () => {
    expect.assertions(1);
    log('info', 'information');
    expect((global.console.info as jest.Mock).mock.calls)
      .toMatchInlineSnapshot(`
      Array [
        Array [
          "[36minformation[39m",
        ],
      ]
    `);
  });

  it('can log error messages', () => {
    expect.assertions(1);
    log('error', 'error message');
    expect((global.console.error as jest.Mock).mock.calls)
      .toMatchInlineSnapshot(`
      Array [
        Array [
          "[37m[41m[1merror message[22m[49m[39m",
        ],
      ]
    `);
  });

  it('can log warning messages', () => {
    expect.assertions(1);
    log('warn', 'warn message');
    expect((global.console.warn as jest.Mock).mock.calls)
      .toMatchInlineSnapshot(`
      Array [
        Array [
          "[33m[1mwarn message[22m[39m",
        ],
      ]
    `);
  });

  it('can write other types of messages', () => {
    expect.assertions(1);
    log('success', 'success message');
    expect((global.console.log as jest.Mock).mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "[32msuccess message[39m",
        ],
      ]
    `);
  });
});
