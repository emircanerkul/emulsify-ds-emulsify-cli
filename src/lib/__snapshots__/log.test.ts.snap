// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`log can log error messages 1`] = `
Array [
  Array [
    "[31m[1merror message[22m[39m",
  ],
]
`;

exports[`log can log warning messages 1`] = `
Array [
  Array [
    "[33m[1mwarn message[22m[39m",
  ],
]
`;

exports[`log can write other types of messages 1`] = `
Array [
  Array [
    "[32msuccess message[39m",
  ],
]
`;

exports[`log exits with the given code if one is provided 1`] = `
Array [
  Array [
    1,
  ],
]
`;
